### PR TITLE
Avoid preprocessor warnings

### PR DESCRIPTION
**Pull request by dankm**
Use ifdef for the `__LP64__` macros. The previous syntax was correct but compilers are starting to warn on undefined preprocessor macros.

### CODE REVIEW

**Review by davidchisnall — APPROVED** [review]

**Comment by DimitryAndric** [comment]
I can't push this, as the repository refuses direct pushes to master. I also can't merge the pull request, since that UI is disabled for me. @davidchisnall please push the button :)

**Comment by dankm** [comment]
I suspect it wants to run the CI checks before it'll allow that. It also looks like kicking off the CI checks is a manual process?

**Comment by davidchisnall** [comment]
CI is currently broken.  https://github.com/libcxxrt/libcxxrt/pull/28 should fix it, but needs @emaste to push a button.

**Comment by dankm** [comment]
Makes sense. Once that's merged will I need to rebase this change, or do the CI checks from the tip master run?

**Comment by davidchisnall** [comment]
Rebased and running CI now.

**Comment by davidchisnall** [comment]
It looks as if something in this PR is changing the number of tests that run between the system runtime and the newly built one, but only on FreeBSD 15 and only with release configurations.  I can't see anything that looks like it would do that.  @DimitryAndric, any thoughts?

**Comment by DimitryAndric** [comment]
> It looks as if something in this PR is changing the number of tests that run between the system runtime and the newly built one, but only on FreeBSD 15 and only with release configurations. I can't see anything that looks like it would do that. @DimitryAndric, any thoughts?

I can't compile it here, due to a mismatch between libc++ 18's declaration and libcxxrt's:

```text
[ 81% 29/32] /usr/bin/c++  -I/share/dim/src/libcxxrt/src -O3 -DNDEBUG -std=gnu++11 -MD -MT test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o -MF test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o.d -o test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o -c /share/dim/src/libcxxrt/test/test_exception.cc
FAILED: test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o
/usr/bin/c++  -I/share/dim/src/libcxxrt/src -O3 -DNDEBUG -std=gnu++11 -MD -MT test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o -MF test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o.d -o test/CMakeFiles/cxxrt-test-shared.dir/test_exception.cc.o -c /share/dim/src/libcxxrt/test/test_exception.cc
/share/dim/src/libcxxrt/test/test_exception.cc:337:19: error: functions that differ only in their return type cannot be overloaded
  337 |         __cxa_exception *__cxa_init_primary_exception(
      |         ~~~~~~~~~~~~~~~~~^
/usr/include/c++/v1/__exception/exception_ptr.h:36:47: note: previous declaration is here
   36 | _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
      |                              ~~~~~~~~~~~~~~~~ ^
1 error generated.
[ 84% 29/32] /usr/bin/c++  -I/share/dim/src/libcxxrt/src -O3 -DNDEBUG -std=gnu++11 -MD -MT test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -MF test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o.d -o test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -c /share/dim/src/libcxxrt/test/test_exception.cc
FAILED: test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o
/usr/bin/c++  -I/share/dim/src/libcxxrt/src -O3 -DNDEBUG -std=gnu++11 -MD -MT test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -MF test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o.d -o test/CMakeFiles/cxxrt-test-static.dir/test_exception.cc.o -c /share/dim/src/libcxxrt/test/test_exception.cc
/share/dim/src/libcxxrt/test/test_exception.cc:337:19: error: functions that differ only in their return type cannot be overloaded
  337 |         __cxa_exception *__cxa_init_primary_exception(
      |         ~~~~~~~~~~~~~~~~~^
/usr/include/c++/v1/__exception/exception_ptr.h:36:47: note: previous declaration is here
   36 | _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
      |                              ~~~~~~~~~~~~~~~~ ^
1 error generated.
```

I will have to locally apply a fix to get around that.

**Comment by DimitryAndric** [comment]
I think I don't understand the tests very well. There is a file `expected_output.log` but I have no idea how it is generated.

I can only say that on a relatively fresh 15-CURRENT test machine, all tests succeed, and it seems that it does 71 tests. I don't really see where that count originates.

**Comment by davidchisnall** [comment]
> I think I don't understand the tests very well. There is a file `expected_output.log` but I have no idea how it is generated.
> 
> I can only say that on a relatively fresh 15-CURRENT test machine, all tests succeed, and it seems that it does 71 tests. I don't really see where that count originates.

The expected output file is generated by running the tests built against the system C++ environment. The goal is to ensure that the test have the same behaviour with libcxxrt and whatever it’s replacing (on FreeBSD, an older version of libcxxrt).

**Comment by DimitryAndric** [comment]
FWIW I don't get any test failures here on `15.0-CURRENT #0 main-n267507-a61d2c7fbd3c` in a fresh VM. I have no idea what version Cirrus CI is running, that information seems impossible to find.

Maybe the Cirrus CI stuff is running an older or newer snapshot? That could be the only reason for the tests to fail, I guess.

What would also help is if the `ctest` command was run with `--verbose`, as at the moment the output is very terse, and just shouting "FAIL!" does not give you any useful information.

**Comment by davidchisnall** [comment]
Cirrus is running whatever re@ ships for Google Compute Engine images.

**Comment by davidchisnall** [comment]
Looks like CI failed on this after the merge.

**Comment by DimitryAndric** [comment]
And again, I have no idea what is going wrong there :)

**Comment by DimitryAndric** [comment]
I think I finally found the culprit: threads being unpredictable in `test/test_guard.cc`. When I run the tests with `-v` so all output of all `TEST()` macros is shown, I see:

```diff
--- cxxrt-test-static-output.log        2024-02-05 13:59:25.196518000 +0100
+++ expected_output.log 2024-02-05 13:59:20.874510000 +0100
@@ -13,6 +13,7 @@
 Test passed: /share/dim/src/libcxxrt/test/test_guard.cc:60: Static initialized
 Test passed: /share/dim/src/libcxxrt/test/test_guard.cc:61: Each static only initialized once
 Test passed: /share/dim/src/libcxxrt/test/test_guard.cc:52: Field correctly initialised
+Test passed: /share/dim/src/libcxxrt/test/test_guard.cc:52: Field correctly initialised
 Test passed: /share/dim/src/libcxxrt/test/test_guard.cc:64: Two threads both tried to initialise a static
 Test passed: /share/dim/src/libcxxrt/test/test_exception.cc:132: Caught int
 Test passed: /share/dim/src/libcxxrt/test/test_exception.cc:336: Cleanup ran correctly
```

What appears to happen is that line 52 (`TEST(s.field == 1, "Field correctly initialised");`) is always succeeding, so `s.field` is indeed 1, but sometimes the line is hit more than once! And then of course the total number of successful tests is 1 higher than before.

I think maybe the `TEST()` macro should not be called inside that function, or alternatively it should be a variant which does not increase the global successful counter?

On the other hand, do we ever expect this line to be visited twice?

**Comment by davidchisnall** [comment]
Thanks!  I think the problem is if second thread is not scheduled before the tests complete.

I think the right fix is to move the declaration of `thr` on line 37 out to the file scope and add a `pthread_join` on line 63.  That will ensure that both threads finish the test checks that they didn't race badly.

**Pull request merged**
